libraries/core/src/topics.rs

```diff
@@ -94,3 +86,125 @@ impl Display for DataflowId {
         }
     }
 }
+
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub struct DataflowResult {
+    pub uuid: Uuid,
+    pub timestamp: uhlc::Timestamp,
+    pub node_results: BTreeMap<NodeId, Result<(), NodeError>>,
+}
+
+impl DataflowResult {
+    pub fn ok_empty(uuid: Uuid, timestamp: uhlc::Timestamp) -> Self {
+        Self {
+            uuid,
+            timestamp,
+            node_results: Default::default(),
+        }
+    }
+
+    pub fn is_ok(&self) -> bool {
+        self.node_results.values().all(|r| r.is_ok())
+    }
+}
+
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub struct DataflowDaemonResult {
+    pub timestamp: uhlc::Timestamp,
+    pub node_results: BTreeMap<NodeId, Result<(), NodeError>>,
+}
+
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub struct NodeError {
+    pub timestamp: uhlc::Timestamp,
+    pub cause: NodeErrorCause,
+    pub exit_status: NodeExitStatus,
+}
+
+impl std::fmt::Display for NodeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self.exit_status {
+            NodeExitStatus::Success => write!(f, "<success>"),
+            NodeExitStatus::IoError(err) => write!(f, "I/O error while reading exit status: {err}"),
+            NodeExitStatus::ExitCode(code) => write!(f, "exited with code: {code}"),
+            NodeExitStatus::Signal(signal) => {
+                let signal_str: Cow<_> = match signal {
+                    1 => "SIGHUP".into(),
+                    2 => "SIGINT".into(),
+                    3 => "SIGQUIT".into(),
+                    4 => "SIGILL".into(),
+                    6 => "SIGABRT".into(),
+                    8 => "SIGFPE".into(),
+                    9 => "SIGKILL".into(),
+                    11 => "SIGSEGV".into(),
+                    13 => "SIGPIPE".into(),
+                    14 => "SIGALRM".into(),
+                    15 => "SIGTERM".into(),
+                    22 => "SIGABRT".into(),
+                    23 => "NSIG".into(),
+                    other => other.to_string().into(),
+                };
+                write!(f, "exited because of signal {signal_str}")
+            }
+            NodeExitStatus::Unknown => write!(f, "unknown exit status"),
+        }?;
+
+        match &self.cause {
+            NodeErrorCause::Cascading { caused_by_node } => write!(
+                f,
+                "\n\nThis error occurred because node `{caused_by_node}` exited before connecting to dora."
+            )?,
+            NodeErrorCause::Other { stderr } if stderr.is_empty() => {}
+            NodeErrorCause::Other { stderr } => {
+                let line: &str = "---------------------------------------------------------------------------------";
+                write!(f, "\n\nStderr output:\n{line}\n{stderr}\n[...]\n{line}\n")?
```

Comment:
The [...] seems to point out that there is more to the error message that there is.

Could we remove it?

Also:

Can we make:
```bash
Node `idefics2` failed: exited with code: 1

Stderr output:
```

Be one line instead of 3 so that it looks like:

```bash
- Node 018ffe7b-39d7-725a-a812-ac9c220a11c8/plot exited with code 1 with error (stderr):
```

Just as in #542

Reply:
> The [...] seems to point out that there is more to the error message that there is.
> 
> Could we remove it?

Parts of the the error message might be cut off. I thought that it might be less confusing with a `[...]` marker in that case (but the marker should be at the beggining, not the end). For example, consider the following error message: `"some\nmultiline\nerror\nmessage\nwith\nlots\nof\nlines\nand\nstuff\nand\nthings"`. It appears like this in the terminal:

```
Stderr output:
---------------------------------------------------------------------------------
[...]
message
with
lots
of
lines
and
stuff
and
things

---------------------------------------------------------------------------------
````


Without the leading `[...]` it's not obvious that relevant parts of the error message might be omitted.



> ```shell
> [...] Node 018ffe7b-39d7-725a-a812-ac9c220a11c8/plot exited [...]
> ```

I don't think that we need to repeat the dataflow UUID again. It's always the same and it just makes the node ID more difficult to read.

Reply:
> Can we make:
> 
> ```shell
> Node `idefics2` failed: exited with code: 1
> 
> Stderr output:
> ```
> Be one line instead of 3

You mean like this:

```
- Node `rust-node` failed: exited with code 1 with stderr output:
---------------------------------------------------------------------------------
[...]
some
multiline
error
message
---------------------------------------------------------------------------------
```

I think it's easier to read without the leading `- ` before `Node`, provided that we keep the lines before and after the stderr output.

I like the lines around stderr because it makes things easier to read when multiple nodes fail. For example, compare:

```
Dataflow 01903a8d-e26e-74f7-8cff-5edc90c1de0b failed:
- Node `rust-node` failed: exited with code 1 with stderr output:
[...]
Traceback (most recent call last):
  File "main.py", line 9, in <module>
    do_stuff()
  File "main.py", line 5, in do_stuff
    raise Exception("test exception")
Exception: test exception

- Node `rust-sink` failed: exited with code 1 with stderr output:
[...]
Error: Failed to read instrs from ./path/to/instrs.json

Caused by:
No such file or directory (os error 2)
```

with:

```
Dataflow 01903a8f-e07f-7136-aad3-214da35d7158 failed:
Node `rust-node` failed: exited with code 1 with stderr output:
---------------------------------------------------------------------------------
[...]
Traceback (most recent call last):
  File "main.py", line 9, in <module>
    do_stuff()
  File "main.py", line 5, in do_stuff
    raise Exception("test exception")
Exception: test exception
---------------------------------------------------------------------------------

Node `rust-sink` failed: exited with code 1 with stderr output:
---------------------------------------------------------------------------------
[...]
Error: Failed to read instrs from ./path/to/instrs.json

Caused by:
No such file or directory (os error 2)
---------------------------------------------------------------------------------
```

To me, the second example seems much more readable.

I like the variant with more spacing even more:

```
Dataflow 01903a92-8c73-770e-b2ca-840dd8c60d0b failed:

Node `rust-node` failed: exited with code 1

Stderr output:
---------------------------------------------------------------------------------
[...]
Traceback (most recent call last):
  File "main.py", line 9, in <module>
    do_stuff()
  File "main.py", line 5, in do_stuff
    raise Exception("test exception")
Exception: test exception
---------------------------------------------------------------------------------

Node `rust-sink` failed: exited with code 1

Stderr output:
---------------------------------------------------------------------------------
[...]
Error: Failed to read instrs from ./path/to/instrs.json

Caused by:
No such file or directory (os error 2)
---------------------------------------------------------------------------------
```

Reply:
I think that the current version:

```bash
019049ab-9473-7c5c-a4b4-d19513e219b5
Dataflow 019049ab-9473-7c5c-a4b4-d19513e219b5 failed:

Node `plot` failed: exited with code 1 with stderr output:
---------------------------------------------------------------------------------
[...]
Traceback (most recent call last):
  File "/home/peter/Documents/work/dora/examples/python-dataflow/plot.py", line 85, in <module>
    assert False, "False"
AssertionError: False
---------------------------------------------------------------------------------

Node `object_detection` failed: exited with code 1 with stderr output:
---------------------------------------------------------------------------------
[...]
Traceback (most recent call last):
  File "/home/peter/Documents/work/dora/examples/python-dataflow/object_detection.py", line 16, in <module>
    assert False, "False"
AssertionError: False
---------------------------------------------------------------------------------

```
Looks good.

I'm slightly confused as, it seems that we are showing all stderr. Do we have a slicing of `stderr` somewhere?

If not, i would not put the `[...]` if that is referencing something like logs.

Reply:
Yes, there is some slicing at the daemon level. But we don't need to show the `[...]` if the stderr buffer is not full (because no slicing happened then). I pushed 5bda3338 to implement this.